daemons/controld/pacemaker-controld.c

```diff
@@ -126,24 +117,36 @@ main(int argc, char **argv)
         fprintf(stderr,
                 "ERROR: Bad permissions on " PE_STATE_DIR " (see logs for details)\n");
         fflush(stderr);
-        return CRM_EX_FATAL;
+        exit_code = CRM_EX_FATAL;
+        goto done;
 
     } else if (pcmk__daemon_can_write(CRM_CONFIG_DIR, NULL) == FALSE) {
         crm_err("Terminating due to bad permissions on " CRM_CONFIG_DIR);
         fprintf(stderr,
                 "ERROR: Bad permissions on " CRM_CONFIG_DIR " (see logs for details)\n");
         fflush(stderr);
-        return CRM_EX_FATAL;
+        exit_code = CRM_EX_FATAL;
+        goto done;
     }
 
     if (pcmk__log_output_new(&(controld_globals.logger_out)) != pcmk_rc_ok) {
-        return CRM_EX_FATAL;
+        exit_code = CRM_EX_FATAL;
+        goto done;
     }
 
     pcmk__output_set_log_level(controld_globals.logger_out, LOG_TRACE);
 
-    crmd_init();
-    return 0; // not reachable
+done:
+    g_strfreev(processed_args);
+    pcmk__free_arg_context(context);
+
+    pcmk__output_and_clear_error(error, NULL);
+
+    if ((exit_code == CRM_EX_OK) && initialize) {
+        // Does not return
+        crmd_init();
+    }
+    crm_exit(exit_code);
```

Comment:
I noticed we free everything else at the end (here and in other daemons) but never unref the mainloop. I don't think `mainloop_cleanup()` in `crm_exit()` does so either. Practically it doesn't matter since we're exiting, just seems odd